pkg/queue/instance.go

```diff
@@ -46,14 +47,16 @@ type Instance interface {
 
 type queueImpl struct {
 	delay     time.Duration
-	tasks     []Task
+	tasks     []*Task
```

Comment:
Do we need a pointer to the Task? Task is already a function which is basically a pointer

Reply:
I think this can be simplified by not storing a map at all. Or maybe getting rid of `metrics` entirely. Something like this:

```diff
diff --git a/pkg/queue/instance.go b/pkg/queue/instance.go
index ea29930966..10854e6824 100644
--- a/pkg/queue/instance.go
+++ b/pkg/queue/instance.go
@@ -47,7 +47,7 @@ type Instance interface {
 
 type queueImpl struct {
 	delay     time.Duration
-	tasks     []*Task
+	tasks     []queueTask
 	cond      *sync.Cond
 	closing   bool
 	closed    chan struct{}
@@ -70,7 +70,7 @@ func NewQueueWithID(errorDelay time.Duration, name string) Instance {
 	}
 	return &queueImpl{
 		delay:       errorDelay,
-		tasks:       make([]*Task, 0),
+		tasks:       make([]queueTask, 0),
 		closing:     false,
 		closed:      make(chan struct{}),
 		closeOnce:   &sync.Once{},
@@ -81,11 +81,17 @@ func NewQueueWithID(errorDelay time.Duration, name string) Instance {
 	}
 }
 
+type queueTask struct {
+	task        Task
+	enqueueTime time.Time
+	startTime time.Time
+}
+
 func (q *queueImpl) Push(item Task) {
 	q.cond.L.Lock()
 	defer q.cond.L.Unlock()
 	if !q.closing {
-		q.tasks = append(q.tasks, &item)
+		q.tasks = append(q.tasks, queueTask{task: item, enqueueTime: time.Now()})
 		if q.metrics != nil {
 			q.metrics.add(&item)
 		}
@@ -100,7 +106,7 @@ func (q *queueImpl) Closed() <-chan struct{} {
 
 // get blocks until it can return a task to be processed. If shutdown = true,
 // the processing go routine should stop.
-func (q *queueImpl) get() (task *Task, shutdown bool) {
+func (q *queueImpl) get() (task queueTask, shutdown bool) {
 	q.cond.L.Lock()
 	defer q.cond.L.Unlock()
 	// wait for closing to be set, or a task to be pushed
@@ -110,15 +116,15 @@ func (q *queueImpl) get() (task *Task, shutdown bool) {
 
 	if q.closing && len(q.tasks) == 0 {
 		// We must be shutting down.
-		return nil, true
+		return queueTask{}, true
 	}
 	task = q.tasks[0]
+	task.startTime = time.Now()
 	// Slicing will not free the underlying elements of the array, so explicitly clear them out here
-	q.tasks[0] = nil
+	q.tasks[0] = queueTask{}
 	q.tasks = q.tasks[1:]
-	if q.metrics != nil {
-		q.metrics.get(task)
-	}
+
+	q.metrics.depth.RecordInt(int64(len(q.tasks)))
 
 	return task, false
 }
@@ -131,16 +137,14 @@ func (q *queueImpl) processNextItem() bool {
 	}
 
 	// Run the task.
-	if err := (*task)(); err != nil {
+	if err := task.task(); err != nil {
 		delay := q.delay
 		log.Infof("Work item handle failed (%v), retry after delay %v", err, delay)
 		time.AfterFunc(delay, func() {
-			q.Push(*task)
+			q.Push(task.task)
 		})
 	}
-	if q.metrics != nil {
-		q.metrics.done(task)
-	}
+	q.metrics.workDuration.Record(time.Since(task.startTime).Seconds())
 
 	return true
 }
```

With metrics we are basically keeping all the same info twice. This may be higher cost and easy to make mistakes where they fall out of sync. Instead we can just pass around the timings? WDYT?

Reply:
Ive been following along on this a bit. I made a PR for some benchmarks. Basically the existing benchmark doesn't work well anymore with this change because of the way its setup. 

It makes a new queue over and over again which isnt really what we want to test 
```
	for n := 0; n < b.N; n++ {
		q := NewQueue(1 * time.Microsecond)
```

https://github.com/ra-grover/istio/pull/1

```
go test -bench=. -count 1 -benchmem
2023-07-06T21:53:46.190758Z	info	Work item handle failed (fake error), retry after delay 1µs
goos: darwin
goarch: arm64
pkg: istio.io/istio/pkg/queue
BenchmarkQueue-10                         	     766	   1366433 ns/op	  630649 B/op	   18114 allocs/op
BenchmarkMetricsQueue-10                  	  685503	      1694 ns/op	     560 B/op	      19 allocs/op
BenchmarkMetricsQueueDisabled-10          	 2983182	       405.9 ns/op	      48 B/op	       4 allocs/op
BenchmarkMetricsQueueAdd-10               	 4464358	       267.9 ns/op	     106 B/op	       3 allocs/op
BenchmarkMetricsQueueInc-10               	 4663022	       257.1 ns/op	      74 B/op	       2 allocs/op
BenchmarkMetricsQueueRec-10               	 4460540	       268.8 ns/op	     106 B/op	       3 allocs/op
BenchmarkMetricsQueueSinceInSeconds-10    	36536446	        32.18 ns/op	       0 B/op	       0 allocs/op
BenchmarkMetricsQueueGet-10               	 1609574	       753.7 ns/op	     320 B/op	      11 allocs/op
PASS
ok  	istio.io/istio/pkg/queue	12.103s
```

I'm not sure how much value the others are but these 2 show the real impact of this change and would love to see how much impact taking the mapping out has. 
```
BenchmarkMetricsQueue-10                  	  685503	      1694 ns/op	     560 B/op	      19 allocs/op
BenchmarkMetricsQueueDisabled-10          	 2983182	       405.9 ns/op	      48 B/op	       4 allocs/op
```